### PR TITLE
ContextRestoringExecutor Hotspots during beforeTask and endTask #1800

### PR DESCRIPTION
**Pull request by jleach4**
Screen Shot 2022-07-22 at 11 17 45 AM

Screen Shot 2022-07-22 at 11 17 06 AM

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/ContextRestoringExecutor.java [hunk 1]
```diff
@@ -56,15 +56,14 @@ public Map<String, String> getMdcContext() {
...
-            MDC.put(entry.getKey(), entry.getValue());
-        }
+        MDC.setContextMap(mdcContext);
     }
 
     static void clearMdc(@Nonnull Map<String, String> mdcContext) {
+        Map<String, String> map = MDC.getMDCAdapter().getCopyOfContextMap();
         for (String key : mdcContext.keySet()) {
-            MDC.remove(key);
+            map.remove(key);
```
Comment (jjezra) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/ContextRestoringExecutor.java:65]: Would it be possible/efficient to use something like: 
```
MDC.keySet().removeAll(mdcContext.keySet());
```

Comment (jleach4) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/ContextRestoringExecutor.java:65]: no

Comment (jleach4) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/ContextRestoringExecutor.java:65]: It is a weird implementation...

**Review by tian-yizuo — APPROVED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/ContextRestoringExecutor.java (hunk 1 above)
Comment (jjezra) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/ContextRestoringExecutor.java:65]: https://stackoverflow.com/questions/17675804/remove-multiple-keys-from-map-in-efficient-way

Comment (MMcM) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/ContextRestoringExecutor.java:65]: This is the limitation of what SLF4J 1.7 can do: https://github.com/qos-ch/slf4j/blob/e9ee55cca93c2bf26f14482a9bdf961c750d2a56/slf4j-api/src/main/java/org/slf4j/spi/MDCAdapter.java

This code is from f5e1b6d23fa18f1c9fa9bfb24ce01ce6cdc758d5 (2017-04-12), when I believe it was still log4j 1. There was some discussion then about advocating for a better API, but nothing came of it. Nothing much has changed in SLF4J 2.

**Review by MMcM — APPROVED** [review]

**Review by jjezra — APPROVED** [review]

**Review by alecgrieser — APPROVED** [review]

**Comment by alecgrieser** [comment]
It might be a good idea to add a release note (probably under the "performance" section) so we know which RL release this improvement ends up in, but otherwise, LGTM

**Pull request merged**